Mathlib/Combinatorics/Optimization/ValuedCSP.lean

```diff
@@ -0,0 +1,79 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.Algebra.Order.Monoid.Defs
+import Mathlib.Data.Fin.VecNotation
+
+/-!
+
+# General-Valued Constraint Satisfaction Problems
+
+General-Valued CSP is a very broad class of problems in discrete optimization.
+General-Valued CSP subsumes Min-Cost-Hom (including 3-SAT for example) and Finite-Valued CSP.
+
+## Main definitions
+* `ValuedCspTemplate`: A VCSP template; fixes a domain, a codomain, and allowed cost functions.
+* `ValuedCspTerm`: One summand in a VCSP instance; calls a concrete function from given template.
+* `ValuedCspTerm.evalSolution`: An evaluation of the VCSP term for given solution.
+* `ValuedCspInstance`: An instance of a VCSP problem over given template.
+* `ValuedCspInstance.evalSolution`: An evaluation of the VCSP instance for given solution.
+* `ValuedCspInstance.optimumSolution`: Is given solution a minimum of the VCSP instance?
+
+## References
+* [D. A. Cohen, M. C. Cooper, P. Creed, P. G. Jeavons, S. Živný,
+   *An Algebraic Theory of Complexity for Discrete Optimisation*][cohen2012]
+
+-/
+
+def n1aryOfUnary {α β : Type*} (f : α → β) : (Fin 1 → α) → β :=
+  fun a => f (a 0)
+
+def n2aryOfBinary {α β : Type*} (f : α → α → β) : (Fin 2 → α) → β :=
+  fun a => f (a 0) (a 1)
+
+/-- A template for a valued CSP problem over a domain `D` with costs in `C`. -/
+@[reducible]
+def ValuedCspTemplate (D C : Type*) [LinearOrderedAddCommMonoid C] :=
+  Set (Σ (n : ℕ), (Fin n → D) → C) -- Cost functions `D^n → C` for any `n`
+
+variable {D C : Type*} [LinearOrderedAddCommMonoid C]
+
+/-- A term in a valued CSP instance over the template `Γ`. -/
+structure ValuedCspTerm (Γ : ValuedCspTemplate D C) (ι : Type*) where
```

Comment:
```suggestion
structure ValuedCspTemplate.Term (Γ : ValuedCspTemplate D C) (ι : Type*) where
```
All of my renamings are suggestions, feel free to not apply them if you want. In this case, this allows you to write `Γ.Term`.